homeassistant/components/graphite.py

```diff
@@ -0,0 +1,131 @@
+    try:
+        port = int(graphite_config.get('port', 2003))
+    except ValueError:
+        _LOGGER.error('Invalid port specified')
+        return False
+
+    _INSTANCE = GraphiteFeeder(hass, host, port, prefix)
+    return True
+
+
+class GraphiteFeeder(threading.Thread):
+    """ Feeds data to graphite. """
+    def __init__(self, hass, host, port, prefix):
+        super(GraphiteFeeder, self).__init__(daemon=True)
+        self._hass = hass
+        self._host = host
+        self._port = port
+        # rstrip any trailing dots in case they think they
+        # need it
+        self._prefix = prefix.rstrip('.')
+        self._queue = queue.Queue()
+        self._quit_object = object()
+
+        hass.bus.listen_once(EVENT_HOMEASSISTANT_START,
+                             self.start_listen)
+        hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP,
+                             self.shutdown)
+        hass.bus.listen(MATCH_ALL, self.event_listener)
+
+    def start_listen(self, event):
+        """ Start event-processing thread. """
+        self.start()
+
+    def shutdown(self, event):
+        """ Tell the thread that we are done.
+
+        This does not block because there is nothing to
+        clean up (and no penalty for killing in-process
+        connections to graphite.
+        """
+        self._queue.put(self._quit_object)
+
+    def event_listener(self, event):
+        """ Queue an event for processing. """
+        self._queue.put(event)
+
+    def _send_to_graphite(self, data):
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        sock.settimeout(10)
+        sock.connect((self._host, self._port))
+        sock.sendall(data.encode('ascii'))
+        sock.send('\n'.encode('ascii'))
+        sock.close()
+
+    def _report_attributes(self, entity_id, state, attributes):
+        now = time.time()
+        things = dict(attributes)
+        if isinstance(state, (float, int)):
+            state = float(state)
+        elif state in (STATE_ON, STATE_OFF):
+            state = float(state == STATE_ON)
+        else:
+            state = None
+        if state is not None:
+            things['state'] = state
+        lines = ['%s.%s.%s %f %i' % (self._prefix,
+                                     entity_id, key.replace(' ', '_'),
+                                     value, now)
+                 for key, value in things.items()
+                 if isinstance(value, (float, int))]
+        if not lines:
+            return
+        _LOGGER.debug('Sending to graphite: %s', lines)
+        try:
+            self._send_to_graphite('\n'.join(lines))
+        except socket.error:
+            _LOGGER.exception('Failed to send data to graphite')
+
+    def run(self):
+        while True:
+            event = self._queue.get()
+            if event == self._quit_object:
+                self._queue.task_done()
+                return
+            elif (event.event_type == EVENT_STATE_CHANGED and
```

Comment:
If you only care about `state_changed` events, consider subscribing to just `state_changed` events.